stablehlo/tests/vhlo_to_version_downgrade_invalid.mlir

```diff
@@ -21,6 +21,18 @@ func.func @custom_call_v2_with_output_operand_alises(%arg0 : tensor<f32>) -> ten
 
 // -----
 
+func.func @op_with_output_operand_alises(%arg0 : tensor<f32>) -> () {
+  // expected-error @+1 {{failed to legalize operation 'vhlo.add' that was explicitly marked illegal}}
+  stablehlo.add %arg0, %arg0 {
+    op_alias_attr = #stablehlo.output_operand_alias<output_tuple_indices = [],
+                                                    operand_index = 0,
+                                                    operand_tuple_indices = []>
```

Comment:
I think I understand why you're doing this (it would be good to test the newly added attribute, but testing stablehlo.custom_call with an output_operand_alias won't achieve much because the test will fail due to a different reason), but I don't think that testing this attribute in isolation is necessary. 

"What's not covered?" in the compatibility RFC says "Unspecced features: We may make backward incompatible changes to features which haven't been specced" and using unregistered attributes counts as an unspecced feature.

Reply:
> using unregistered attributes counts as an unspecced feature

I could see unregistered attributes being a silent compatibility-killer. If anything changes about an unregistered attr, the file won't deserialize. With that, there were two possible approaches:

1. Enforce a disallowance of unregistered attributes. This felt too strict since it seems like there _have_ been use cases for unregistered attrs in the past.
2. Don't worry about the attrs, so long as their values are representable and valid in VHLO. 

I went with approach 2. If we wanted to go with approach 1, I would want to make sure we are explicit that that is an unpsecced feature.

> won't achieve much because the test will fail due to a different reason

Outside of VHLO downgrade errors, this does not fail. Aside from the unregistered operation at the bottom of this file, the file if fully round-trippable through stablehlo-opt.

Reply:
In the meantime, I added a comment to the test to explain that rationale.

Reply:
"I could see unregistered attributes being a silent compatibility-killer". Let's open a ticket to get to the bottom of this. I agree that it's too early to disallow unregistered attributes in VHLO while we have stuff like `mhlo.sharding` and about a handful of other public attributes unspecced, but could we perhaps have an allowlist for now and ban everything that's not there?

"In the meantime, I added a comment to the test to explain that rationale". Understood. After our discussion on a somewhat similar MLIR-HLO PR recently, I have another concern though. Semantically, having an `output_operand_alias` attribute on an `add` doesn't make sense, which doesn't make this a great test in my opinion. Perhaps we can drop this test for now, and once we have that allowlist (in a follow-up PR), we could introduce a more natural negative, e.g. a unit attribute whose name is not on the list?

Reply:
"but could we perhaps have an allowlist for now and ban everything that's not there"
Opened #784. We would need to version this allowlist as well. The alternative is supporting unregistered attrs so long as their values are supported by VHLO, which feels easier to document, use, and maintain.

"Perhaps we can drop this test for now, and once we have that allowlist (in a follow-up PR), we could introduce a more natural negative"
I'd propose the opposite, VHLO needs testing of unregistered attributes, and I'm not sure that allow-list is the proper path forward just yet. I'll change this to an unregistered attr of a custom_call since that makes a little more sense, but this test point has little to do with op semantics, nor the implementation defined semantics of unregistered attributes. This TP is more akin to [mhlo_bytecode_customizations.mlir](https://github.com/tensorflow/mlir-hlo/blob/984b124ecf96d3c0d1d447ffb70c1baca3d9868c/tests/Dialect/mhlo/mhlo_bytecode_customizations.mlir) - nonsense for the sake of testing something else entirely. In this case, we are testing that serialization versioning prevents the legalization of VHLO for a version that would not understand how to deserialize it. If this were to sneak through, even though it is nonsense and may be dropped, it would break deserialization entirely.

Per our discussion on the MLIR-HLO PR, I realized via test cases like this that we are able to add unregistered attributes into `func.func` attributes, which completely violates compatibility currently, re-enforcing the need to fork modularity ops, and have TPs like this for VHLO. While the need is much less for MHLO->StableHLO, I noted that it is _very_ important to test in VHLO, given its use case, and since this _is_ valid (but meaningless) MLIR.

Reply:
Heh - turns out that this attr does sneak through when used as an unregistered attribute of `custom_call` currently. Will push a fix for that shortly.

Reply:
Sounds good.